File: leetcode/binary_search/2838_Maximum_Coins_Heroes_Can_Collect/codes/bs.c
#include <stdio.h>
#include <stdlib.h>

int cmp(const void* a, const void* b) {
    return ((int*)a)[0] - ((int*)b)[0];
}

int upper_bound(int arr[][2], int n, int x) {
    int l = 0, r = n;
    while (l < r) {
        int m = (l + r) / 2;
        if (arr[m][0] <= x) l = m + 1;
        else r = m;
    }
    return l;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
long long* maximumCoins(int* heroes, int heroesSize, int* monsters, int monstersSize, int* coins, int coinsSize, int* returnSize) {
    int (*mon_coin)[2] = malloc(sizeof(int) * 2 * monstersSize);
    for (int i = 0; i < monstersSize; ++i) {
        mon_coin[i][0] = monsters[i];
        mon_coin[i][1] = coins[i];
    }
    qsort(mon_coin, monstersSize, sizeof(mon_coin[0]), cmp);

    long long* prefix = malloc(sizeof(long long) * (monstersSize + 1));
    prefix[0] = 0;
    for (int i = 0; i < monstersSize; ++i) {
        prefix[i + 1] = prefix[i] + mon_coin[i][1];
    }

    long long* ans = malloc(sizeof(long long) * heroesSize);
    for (int i = 0; i < heroesSize; ++i) {
        int idx = upper_bound(mon_coin, monstersSize, heroes[i]);
        ans[i] = prefix[idx];
    }

    free(mon_coin);
    free(prefix);
    *returnSize = heroesSize;
    return ans;
}

File: leetcode/binary_search/2838_Maximum_Coins_Heroes_Can_Collect/codes/bs.py
from typing import List

class Solution:
    def maximumCoins(self, heroes: List[int], monsters: List[int], coins: List[int]) -> List[int]:
        paired = sorted(zip(monsters, coins))
        monsters_sorted = [m for m, _ in paired]
        prefix_sum = [0] * (len(coins) + 1)
        for i in range(len(coins)):
            prefix_sum[i + 1] = prefix_sum[i] + paired[i][1]

        def upper_bound(arr, x):
            left, right = 0, len(arr)
            while left < right:
                mid = (left + right) // 2
                if arr[mid] <= x:
                    left = mid + 1
                else:
                    right = mid
            return left

        result = []
        for h in heroes:
            idx = upper_bound(monsters_sorted, h)
            result.append(prefix_sum[idx])

        return result
